src/persist-client/src/internal/trace.rs

```diff
@@ -1276,60 +1271,40 @@ impl<T: Timestamp + Lattice> MergeState<T> {
         compaction_frontier: Option<AntichainRef<T>>,
     ) -> MergeState<T> {
-        let variant = match (batch1, batch2) {
+        match (batch1, batch2) {
             (Some(batch1), Some(batch2)) => {
                 assert!(batch1.upper() == batch2.lower());
                 let begin_merge = SpineBatch::begin_merge(&batch1, &batch2, compaction_frontier);
-                MergeVariant::InProgress(batch1, batch2, begin_merge)
+                MergeState::Double(MergeVariant::InProgress(batch1, batch2, begin_merge))
             }
-            (None, Some(x)) => MergeVariant::Complete(Some(x)),
-            (Some(x), None) => MergeVariant::Complete(Some(x)),
-            (None, None) => MergeVariant::Complete(None),
-        };
-
-        MergeState::Double(variant)
+            (None, Some(x)) => MergeState::Double(MergeVariant::Complete(x)),
+            (Some(x), None) => MergeState::Double(MergeVariant::Complete(x)),
+            (None, None) => MergeState::Single(None),
+        }
     }
 }
 
 #[derive(Debug, Clone)]
 enum MergeVariant<T> {
     /// Describes an actual in-progress merge between two non-trivial batches.
     InProgress(SpineBatch<T>, SpineBatch<T>, FuelingMerge<T>),
-    /// A merge that requires no further work. May or may not represent a
-    /// non-trivial batch.
-    Complete(Option<SpineBatch<T>>),
+    /// A merge that requires no further work.
+    Complete(SpineBatch<T>),
 }
 
 impl<T: Timestamp + Lattice> MergeVariant<T> {
-    /// Completes and extracts the batch, unless structurally empty.
-    ///
-    /// The result is either `None`, for structurally empty batches, or a batch
-    /// and optionally input batches from which it derived.
-    fn complete(mut self, log: &mut SpineLog<'_, T>) -> Option<SpineBatch<T>> {
-        let mut fuel = isize::max_value();
-        self.work(&mut fuel, log);
-        if let MergeVariant::Complete(batch) = self {
-            batch
-        } else {
-            panic!("Failed to complete a merge!");
+    /// Completes and extracts the batch.
+    fn complete(self, log: &mut SpineLog<'_, T>) -> SpineBatch<T> {
+        match self {
+            MergeVariant::InProgress(a, b, merge) => merge.done(a, b, log),
+            MergeVariant::Complete(batch) => batch,
         }
     }
 
     /// Applies some amount of work, potentially completing the merge.
-    ///
-    /// In case the work completes, the source batches are returned. This allows
-    /// the caller to manage the released resources.
-    fn work(&mut self, fuel: &mut isize, log: &mut SpineLog<'_, T>) {
-        let variant = std::mem::replace(self, MergeVariant::Complete(None));
-        if let MergeVariant::InProgress(b1, b2, mut merge) = variant {
+    fn work(&mut self, fuel: &mut isize) {
+        if let MergeVariant::InProgress(b1, b2, merge) = self {
             merge.work(&b1, &b2, fuel);
-            if *fuel > 0 {
-                *self = MergeVariant::Complete(Some(merge.done(b1, b2, log)));
```

Comment:
Interesting! Not sure I remember enough of the context here to convince myself that it's safe to remove the behavior of `work` automatically converting the batch to a `Complete` when it's been fully fueled. Mind walking me through it?

Reply:
Sure! Looking at: https://github.com/MaterializeInc/materialize/pull/26121/commits/3e94accdf761418bd31ee941d82a560f26d7f980#diff-be562d3f5bcccdd4feabf92e78906997d7addec7c9d7aacd446e8a2df0edbf26L967-L982
- We first apply work. This is the only place that work is applied, and that's the only place an in-progress merge could become complete.
- We then check if the merge is complete, and if so, extract the completed batch.

So a nontrivial `MergeVariant::Complete` was only ever very short-lived in practice: created on line 967 and destroyed on line 980. We do need to adjust the definition of `is_complete` (which is also only ever called here) but otherwise nothing else is affected by deferring the call to `done` a bit.

(Why mess with this at all? Removing the `MergeVariant::Complete(None)` case meant I couldn't keep the `std::mem::replace` above, and this was the simplest change I could find that made the ownership work.)